### PR TITLE
[splash] add warning for splash screen if visible for too long

### PR DESCRIPTION
**Pull request by ajsmth**
@bbarthec hoping to get some feedback on this one before I dive into a solution!

# Why

Users can sometimes prevent hiding splash screen and forget to actually hide it, or leave AppLoading rendered and other such things. We may want to warn people about this in development.

# How

After taking a naive stab at implementing this in JS, it's clear that any solution will be a little bit more complex. 

What is currently preventing this from being fixed in just JS is that the splash screen is rendered natively on top of the React application, so any kind of `console.warn` is rendered underneath this view (e.g not visible)

I think the most ideal solution will be to render the same warnings/error UI that developers are used to. After some investigation it's clear that the warning and error logs are actually a React component that is rendered on top of other React application code in the `AppContainer` that wraps every app: https://github.com/facebook/react-native/blob/master/Libraries/ReactNative/AppContainer.js#L78-L88

In order to get similar functionality I am thinking that we should embed the same React component in the SplashScreen view - I've done preliminary tests in JS where I removed the default LogBox component rendered from `AppContainer` and instead rendered my own, and the logs/warnings worked just the same.

Another option would be to move the splash screen to be a React view that is a sibling of LogBox and the RN app - it looks like `renderApplication` has the ability to wrap React apps with our own wrapper: 

https://github.com/facebook/react-native/blob/master/Libraries/ReactNative/AppContainer.js#L103-L115

I think the first option is more closely aligned with the current implementation, where the splash screen lives outside of the React application - but I'm not sure if it's possible to embed the LogBox component in that screen or how complex that would be. 

The second option is quite a bit of a change in the implementation, but would use the same LogBox that the rest of the React application is using.

I would love to have some feedback and know if perhaps I missed / overlooked something!

### CODE REVIEW

**Comment by bbarthec** [comment]
@ajsmth, I'm really impressed by the amount of research you've performed here! Good work already 😁 💪

I'm afraid option 2 is rather out of question 😞 
> Another option would be to move the splash screen to be a React view that is a sibling of LogBox and the RN app

That's because when the RN app starts the launch flow looks roughly like this:
```
(visible until some rootView is mounted)   (empty at this point)     (takes a while)    (at some point sth is appearing on the screen)
nativeSplashScreen                       -> rootView mounted      -> JS starts       -> some RN render happens
                                         └> ExpoSplashScreen mounted                 -> splash autohide
```

I wonder about option 1, because `SplashScreen` is completely outside the React views tree 🤔 
Is it possible to embed some `React` view inside `SplashScreen` views tree?

**Comment by brentvatne** [comment]
i think we could leverage an API like toast on android to display a message here (we do this in some other places also, like https://github.com/expo/expo/pull/12307). i'm not sure what the best option on the ios side would be - we could pull in some library for this or write our own simple toast api. regardless of what we choose, displaying a warning from native is becoming a common enough requirement that we may want to decide on a standard way of doing this and think through the problem generally.

**Comment by bbarthec** [comment]
I was thinking about it a while longer and came up with an idea 😃 what about listening/subscribing for possible JS error/warnings and once some of these happen we hide spalsh screen and add another warning about the reason it was hidden? 🤔
We can also add some timer/countdown when to hide splash upon warning and immediately upon error message?

**Comment by ajsmth** [comment]
> I was thinking about it a while longer and came up with an idea 😃 what about listening/subscribing for possible JS error/warnings and once some of these happen we hide spalsh screen and add another warning about the reason it was hidden? 🤔
> We can also add some timer/countdown when to hide splash upon warning and immediately upon error message?

I like this idea, it would solve the issue for SplashScreen - but @brentvatne mentions that this is becoming a common problem for other packages as well - so maybe we should look at a more general solution on the native side as well!

**Comment by bbarthec** [comment]
Yeah, I like both ideas to be honest 😅 
I wonder what are the other cases when we need some native mechanism for showing warning/errors? 🤔 cc @brentvatne 
Depending on answer I Would advocate for either global solution or splash-screen-local solution 😉 
Apart fro that, adding some timer to the SplashScreen seems as a valid idea anyway 💪 
Let's do that! 😁

**Comment by brentvatne** [comment]
@bbarthec - so far we have needed it for expo-network (to deprecate in old sdks) and expo-media-library (to warn about android 11 changes in older sdks)

**Comment by bbarthec** [comment]
@brentvatne, can't we, for both mentioned cases, use regular RN's LogBox? 🤔

**Comment by brentvatne** [comment]
@bbarthec - i'm not sure how we could do this from native, what did you have in mind?

**Comment by bbarthec** [comment]
@brentvatne, can't we use this mechanism based on `RCTLogWarn`/`RCTLogError`: https://github.com/expo/react-native/blob/85f6d8f2b00f72ee18cedef2052471ae16c34e26/React/Base/RCTLog.h#L34-L35 ? 🤔 
and `RNLog.e`/`RNLog.w`: https://github.com/expo/react-native/blob/85f6d8f2b00f72ee18cedef2052471ae16c34e26/ReactAndroid/src/main/java/com/facebook/react/util/RNLog.java#L63-L77 ?
My understanding is that using these will result in showing `LogBox` in RN app with either warning or error screen.

**Comment by ajsmth** [comment]
@bbarthec  Made a first stab at implementing this on iOS - let me know your thoughts!

**Comment by ajsmth** [comment]
Heres a sample with a faster timeout -- also disappears after ~8 seconds of inactivity

https://user-images.githubusercontent.com/40680668/119907678-a9d54480-bf05-11eb-8569-6f5640b73b07.mov

**Comment by brentvatne** [comment]
maybe "Still seeing splash screen?" to "Stuck on splash screen?"

**Comment by ajsmth** [comment]
@brentvatne @bbarthec this should be good for a review now if you have some time!

**Review by brentvatne — CHANGES_REQUESTED** [review]
ui is looking good! i think we need to move this logic in to expo go though because we only want this to apply to apps loaded in expo go in development

File: packages/expo-splash-screen/ios/EXSplashScreen/EXSplashScreenService.m [hunk 1]
```diff
@@ -31,17 +32,15 @@ - (void)showSplashScreenFor:(UIViewController *)viewController
                    failureCallback:^(NSString *message){ UMLogWarn(@"%@", message); }];
 }
 
-- (void)showSplashScreenFor:(UIViewController *)viewController
-   splashScreenViewProvider:(id<EXSplashScreenViewProvider>)splashScreenViewProvider
-            successCallback:(void (^)(void))successCallback
-            failureCallback:(void (^)(NSString * _Nonnull))failureCallback
+
+- (void)showSplashScreenFor:(UIViewController *)viewController splashScreenViewProvider:(id<EXSplashScreenViewProvider>)splashScreenViewProvider successCallback:(void (^)(void))successCallback failureCallback:(void (^)(NSString * _Nonnull))failureCallback
```
Comment (brentvatne) [inline packages/expo-splash-screen/ios/EXSplashScreen/EXSplashScreenService.m:36]: format on separate lines similar to the previous formatting

File: packages/expo-splash-screen/ios/EXSplashScreen/EXSplashScreenService.m [hunk 2]
```diff
@@ -31,17 +32,15 @@ - (void)showSplashScreenFor:(UIViewController *)viewController
...
 {
   if ([self.splashScreenControllers objectForKey:viewController]) {
     return failureCallback(@"'SplashScreen.show' has already been called for given view controller.");
   }
   
-  EXSplashScreenController *splashScreenController = [[EXSplashScreenController alloc] initWithViewController:viewController
-                                                                                     splashScreenViewProvider:splashScreenViewProvider];
+  EXSplashScreenController *splashScreenController = [[EXSplashScreenController alloc] initWithViewController:viewController splashScreenViewProvider:splashScreenViewProvider];
```
Comment (brentvatne) [inline packages/expo-splash-screen/ios/EXSplashScreen/EXSplashScreenService.m:42]: same as above re formatting. we need code formatter / linter set up here :O

File: packages/expo-splash-screen/ios/EXSplashScreen/EXSplashScreenController.m [hunk 3]
```diff
@@ -44,12 +45,14 @@ - (void)showWithCallback:(nullable void(^)(void))successCallback
...
     [rootView addSubview:self.splashScreenView];
     self.splashScreenShown = YES;
+    
     if (successCallback) {
       successCallback();
     }
   }];
 }
 
+
```
Comment (brentvatne) [inline packages/expo-splash-screen/ios/EXSplashScreen/EXSplashScreenController.m:55]:
```suggestion
```

File: packages/expo-splash-screen/android/src/main/java/expo/modules/splashscreen/SplashScreenWarningActivity.kt [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package expo.modules.splashscreen
+
+interface SplashScreenWarningActivity {
```
Comment (brentvatne) [inline packages/expo-splash-screen/android/src/main/java/expo/modules/splashscreen/SplashScreenWarningActivity.kt:3]: i don't think the interface name here fits its definition, given that the interface only consists of `onSplashScreenDismissed` which could be used for a variety of purposes

Comment (ajsmth) [reply packages/expo-splash-screen/android/src/main/java/expo/modules/splashscreen/SplashScreenWarningActivity.kt:3]: agreed!

**Review by Kudo — COMMENTED** [review]
for my personal preference:
it seems our splash screen is based on a real view,
so we may able to customize and even add some text directly inside the view.
this could eliminate the usage of SnackBar or MBProgressHUD.

e.g.
android: derived `SplashScreenView.kt` inside expo go.

ios: `configureSplashScreenView:previousConfiguration:` in `EXManagedAppSplashScreenViewProvider.m`

File: packages/expo-splash-screen/android/src/main/java/expo/modules/splashscreen/SplashScreenListenerActivity.kt [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package expo.modules.splashscreen
+
+interface SplashScreenListenerActivity {
+    fun onSplashScreenDismissed(): Unit
```
[the remark this answers is not in the document]
Comment (ajsmth) [reply packages/expo-splash-screen/android/src/main/java/expo/modules/splashscreen/SplashScreenListenerActivity.kt:4]: I like this idea but I'm not sure I totally follow - the view provider classes don't inherit anything from the controllers (from what I can tell) - we could extend the SplashScreen.show to include a controller as an argument and implement `ManagedAppSplashScreenController.kt` which overrides the onDismissed method - is this sort of what you're getting at?

[the remark this answers is not in the document]
Comment (Kudo) [reply packages/expo-splash-screen/android/src/main/java/expo/modules/splashscreen/SplashScreenListenerActivity.kt:4]: I am sorry to say wrong and you are right, view providers does not inherit from controller.

still think on how we could not avoid introducing additional interface and here is yet another try by the `onViewDetachedFromWindow`.
when splash view is dismissed, the callback should be called.
here is the code snippet (taking `ManagedAppSplashScreenViewProvider.kt` as example)

```diff
--- a/android/expoview/src/main/java/host/exp/exponent/experience/splashscreen/ManagedAppSplashScreenViewProvider.kt
+++ b/android/expoview/src/main/java/host/exp/exponent/experience/splashscreen/ManagedAppSplashScreenViewProvider.kt
@@ -1,9 +1,14 @@
 package host.exp.exponent.experience.splashscreen

 import android.content.Context
+import android.graphics.Color
+import android.os.Handler
+import android.os.Looper
 import android.view.View
 import android.view.animation.AccelerateDecelerateInterpolator
 import android.view.animation.AlphaAnimation
+import android.widget.RelativeLayout
+import android.widget.TextView
 import com.squareup.picasso.Callback
 import com.squareup.picasso.Picasso
 import expo.modules.splashscreen.SplashScreenView
@@ -19,6 +24,8 @@ class ManagedAppSplashScreenViewProvider(
   private var config: ManagedAppSplashScreenConfiguration
 ) : SplashScreenViewProvider {
   private lateinit var splashScreenView: SplashScreenView
+  private val handler = Handler(Looper.getMainLooper())
+  private var timeoutRunnable: Runnable? = null

   companion object {
     private const val TAG: String = "ExperienceSplashScreenManifestBasedResourceProvider"
@@ -27,6 +34,7 @@ class ManagedAppSplashScreenViewProvider(
   override fun createSplashScreenView(context: Context): View {
     splashScreenView = SplashScreenView(context)
     configureSplashScreenView(context, config, null)
+    configureTimeoutHintView(context)
     return splashScreenView
   }

@@ -72,4 +80,34 @@ class ManagedAppSplashScreenViewProvider(
       }
     })
   }
+
+  private fun configureTimeoutHintView(context: Context) {
+    splashScreenView.addOnAttachStateChangeListener(object : View.OnAttachStateChangeListener {
+      override fun onViewAttachedToWindow(v: View?) {
+        if (timeoutRunnable == null) {
+          timeoutRunnable = Runnable { showTimeoutHintView(context) }
+          handler.postDelayed(timeoutRunnable!!, 2000)
+        }
+      }
+
+      override fun onViewDetachedFromWindow(v: View?) {
+        timeoutRunnable?.let { handler.removeCallbacks(it) }
+        timeoutRunnable = null
+      }
+    });
+  }
+
+  private fun showTimeoutHintView(context: Context) {
+    val textView = TextView(context)
+    textView.text = "Timeout!"
+    textView.setTextColor(Color.BLACK)
+
+    val layoutParams = RelativeLayout.LayoutParams(
+            RelativeLayout.LayoutParams.WRAP_CONTENT,
+            RelativeLayout.LayoutParams.WRAP_CONTENT)
+    layoutParams.addRule(RelativeLayout.CENTER_HORIZONTAL, RelativeLayout.TRUE)
+    layoutParams.addRule(RelativeLayout.ALIGN_PARENT_BOTTOM, RelativeLayout.TRUE)
+    layoutParams.bottomMargin = 100
+    splashScreenView.addView(textView, layoutParams)
+  }
 }
```

[the remark this answers is not in the document]
Comment (ajsmth) [reply packages/expo-splash-screen/android/src/main/java/expo/modules/splashscreen/SplashScreenListenerActivity.kt:4]: no problem! I like this approach as well as its completely encapsulated by the view provider, but I don't think it would be the same for iOS, I think we'd have to provide a view controller as part of the view provider to get those view lifecycle events, which looks to be a good amount of refactoring.

another idea would be to have the viewprovider specify its splash screen controller via a `createSplashController` method, this way we could subclass the splash controller and get the behaviour we want. This would also be a bit of refactoring, but I think its the best option - what do you think? @Kudo

[the remark this answers is not in the document]
Comment (ajsmth) [reply packages/expo-splash-screen/android/src/main/java/expo/modules/splashscreen/SplashScreenListenerActivity.kt:4]: I've updated with this implementation on iOS - let me know what you think!

[the remark this answers is not in the document]
Comment (Kudo) [reply packages/expo-splash-screen/android/src/main/java/expo/modules/splashscreen/SplashScreenListenerActivity.kt]: this approach is clever, opening the possibility to extend splash controller and cross-platformed.
thanks for the following up.

File: packages/expo-splash-screen/ios/EXSplashScreen/EXSplashScreenController.m [hunk 6]
```diff
@@ -17,15 +17,15 @@ @interface EXSplashScreenController ()
...
 {
   if (self = [super init]) {
-    _viewController = viewController;
+    _rootView = rootView;
     _autoHideEnabled = YES;
     _splashScreenShown = NO;
     _appContentAppeared = NO;
-    _splashScreenView = [splashScreenViewProvider createSplashScreenView];
+    _splashScreenView = splashScreenView;
+    _splashScreenView.userInteractionEnabled = YES;
```
Comment (Kudo) [inline packages/expo-splash-screen/ios/EXSplashScreen/EXSplashScreenController.m:28]: for normal usage, the splash screen view does not need to be interactive.
how about moving this to `EXManagedAppSplashScreenController`?

**Review by Kudo — APPROVED** [review]
just some nits and mostly looks good to me. 🚀

File: android/expoview/src/main/java/host/exp/exponent/experience/ExperienceActivity.java [hunk 7]
```diff
@@ -415,7 +415,6 @@ protected void showOrReconfigureManagedAppSplashScreen(final RawManifest manifes
       mManagedAppSplashScreenViewProvider.updateSplashScreenViewWithManifest(this, manifest);
     }
   }
-
```
Comment (brentvatne) [inline android/expoview/src/main/java/host/exp/exponent/experience/ExperienceActivity.java:418]: intentional to remove this newline?

File: android/expoview/src/main/java/host/exp/exponent/experience/splashscreen/ManagedAppSplashScreenController.kt [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+
+    override fun hideSplashScreen(successCallback: (hasEffect: Boolean) -> Unit, failureCallback: (reason: String) -> Unit) {
+        super.hideSplashScreen({
+            mRunnable?.let { it1 -> mWarningHandler.removeCallbacks(it1) }
+            successCallback(it)
+        }, failureCallback)
+    }
+
+
```
Comment (brentvatne) [inline android/expoview/src/main/java/host/exp/exponent/experience/splashscreen/ManagedAppSplashScreenController.kt:51]:
```suggestion
```

File: android/expoview/src/main/java/host/exp/exponent/experience/splashscreen/ManagedAppSplashScreenController.kt [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    override fun showSplashScreen(successCallback: () -> Unit) {
+        super.showSplashScreen {
+            startSplashScreenWarningTimer()
+            successCallback()
+        }
+    }
+
```
Comment (brentvatne) [inline android/expoview/src/main/java/host/exp/exponent/experience/splashscreen/ManagedAppSplashScreenController.kt:42]:
```suggestion
```

File: android/expoview/src/main/java/host/exp/exponent/experience/splashscreen/ManagedAppSplashScreenController.kt [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+                val intent = Intent(Intent.ACTION_VIEW, webpage)
+                v.context.startActivity(intent)
+                mSnackbar!!.dismiss()
+            })
+            mSnackbar!!.show()
+        }
+
+        mWarningHandler.postDelayed(mRunnable!!, 20000)
+    }
+
```
Comment (brentvatne) [inline android/expoview/src/main/java/host/exp/exponent/experience/splashscreen/ManagedAppSplashScreenController.kt:34]:
```suggestion
```

File: apps/bare-expo/ios/Podfile.lock [hunk 11]
```diff
@@ -1311,7 +1311,7 @@ SPEC CHECKSUMS:
   RCTTypeSafety: 44982c5c8e43ff4141eb519a8ddc88059acd1f3a
   React: e1c65dd41cb9db13b99f24608e47dd595f28ca9a
   React-callinvoker: 552a6a6bc8b3bb794cf108ad59e5a9e2e3b4fc98
-  React-Core: 9d341e725dc9cd2f49e4c49ad1fc4e8776aa2639
+  React-Core: a42b4deaadbde5a2d8e2dab7a2b874303f733cf4
```
Comment (brentvatne) [inline apps/bare-expo/ios/Podfile.lock:1314]: intentional? seems unrelated to this pr

File: ios/Exponent/Kernel/Views/EXAppViewController.m [hunk 12]
```diff
@@ -71,6 +71,7 @@ @interface EXAppViewController ()
 @property (nonatomic, assign) BOOL isStandalone;
 @property (nonatomic, assign) BOOL isHomeApp;
 
+
```
Comment (brentvatne) [inline ios/Exponent/Kernel/Views/EXAppViewController.m:74]: unnecessary whitespace

File: ios/Exponent/Kernel/Views/EXAppViewController.m [hunk 13]
```diff
@@ -337,7 +338,6 @@ - (void)_showOrReconfigureManagedAppSplashScreen:(EXUpdatesRawManifest *)manifes
   }
   if (!_managedAppSplashScreenViewProvider) {
     _managedAppSplashScreenViewProvider = [[EXManagedAppSplashScreenViewProvider alloc] initWithManifest:manifest];
-
```
Comment (brentvatne) [inline ios/Exponent/Kernel/Views/EXAppViewController.m:340]: intentional?

File: ios/Exponent/Kernel/Views/EXAppViewController.m [hunk 14]
```diff
@@ -375,7 +375,6 @@ - (void)_setLoadingViewStatusIfEnabledFromAppLoader:(EXAppLoader *)appLoader
 - (void)_showSplashScreenWithProvider:(id<EXSplashScreenViewProvider>)provider
 {
   EXSplashScreenService *splashScreenService = (EXSplashScreenService *)[UMModuleRegistryProvider getSingletonModuleForClass:[EXSplashScreenService class]];
-
```
Comment (brentvatne) [inline ios/Exponent/Kernel/Views/EXAppViewController.m:378]: intentional?

File: ios/Exponent/Kernel/Views/Loading/EXManagedAppSplashScreenController.m [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+-(void)startSplashScreenVisibleTimer
+{
+  self.warningTimer = [NSTimer scheduledTimerWithTimeInterval:20.0
+                                                       target:self
+                                                     selector:@selector(showSplashScreenVisibleWarning)
+                                                     userInfo:nil
+                                                      repeats:NO];
+}
+
+-(void)showSplashScreenVisibleWarning
```
Comment (brentvatne) [inline ios/Exponent/Kernel/Views/Loading/EXManagedAppSplashScreenController.m:69]:
```suggestion
- (void)showSplashScreenVisibleWarning
```

File: ios/Exponent/Kernel/Views/Loading/EXManagedAppSplashScreenController.m [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+  [button addTarget:self action:@selector(navigateToFYI) forControlEvents:UIControlEventTouchUpInside];
+
+  _warningHud.customView = button;
+  _warningHud.offset = CGPointMake(0.f, MBProgressMaxOffset);
+  
+  [_warningHud hideAnimated:YES afterDelay:8.f];
+#endif
+}
+
+-(void)navigateToFYI
```
Comment (brentvatne) [inline ios/Exponent/Kernel/Views/Loading/EXManagedAppSplashScreenController.m:85]:
```suggestion
- (void)navigateToFYI
```

File: packages/expo-splash-screen/CHANGELOG.md [hunk 17]
```diff
@@ -5,6 +5,7 @@
 ### 🛠 Breaking changes
 
 ### 🎉 New features
+- Add warning for splash screen if visible for too long ([#12882](https://github.com/expo/expo/pull/12882) by [@ajsmth](https://github.com/ajsmth))
```
Comment (brentvatne) [inline packages/expo-splash-screen/CHANGELOG.md:9]: another little spacing nit - newline should be after header and no extra newline between new feature items in the changelog list

**Review by brentvatne — APPROVED** [review]
some small nits (we really need formatting linting set up here!) but otherwise lg

File: android/expoview/src/main/java/host/exp/exponent/experience/splashscreen/ManagedAppSplashScreenController.kt [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+                val webpage = Uri.parse(url)
+                val intent = Intent(Intent.ACTION_VIEW, webpage)
+                v.context.startActivity(intent)
+                mSnackbar!!.dismiss()
+            })
+            mSnackbar!!.show()
+        }
+
+        mWarningHandler.postDelayed(mRunnable!!, 20000)
+    }
```
Comment (bbarthec) [inline android/expoview/src/main/java/host/exp/exponent/experience/splashscreen/ManagedAppSplashScreenController.kt:33]:
```suggestion
  private fun startSplashScreenWarningTimer() {
    mRunnable = Runnable {
      mSnackbar = Snackbar.make(splashScreenView, "Stuck on splash screen?", Snackbar.LENGTH_LONG).also { snackbar ->
        snackbar.setAction("Info", View.OnClickListener { v ->
          val url = "https://expo.fyi/splash-screen-hanging"
          val webpage = Uri.parse(url)
          val intent = Intent(Intent.ACTION_VIEW, webpage)
          v.context.startActivity(intent)
          snackbar.dismiss()
        })
      }
      snackbar.show()
    }
     
    mWarningHandler.postDelayed(mRunnable!!, 20000)
  }
```

File: ios/Exponent/Kernel/Views/Loading/EXManagedAppSplashScreenController.m [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+#import "EXManagedAppSplashScreenController.h"
+#import "MBProgressHUD.h"
+#import "EXSplashScreenHUDButton.h"
+
+
+@interface EXManagedAppSplashScreenController()
+
+@property (nonatomic, weak) NSTimer *warningTimer;
+@property (nonatomic, weak) MBProgressHUD *warningHud;
+@property (nonatomic, strong) UIView *splashScreenView;
```
Comment (bbarthec) [inline ios/Exponent/Kernel/Views/Loading/EXManagedAppSplashScreenController.m:18]: From what I see the base class `SplashScreenController` is already holding a strong reference to this view 🤔 Not sure about `strong` modifier here 🧐 
Unfortunately, AFAIK, there's no `protected` property in Objective-C to get rid of this altogether 😞

File: ios/Exponent/Kernel/Views/Loading/EXManagedAppSplashScreenController.m [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+
+@end
+
+@implementation EXManagedAppSplashScreenController
+
+- (instancetype)initWithRootView:(UIView *)rootView splashScreenView:(UIView *)splashScreenView
+{
+  self = [super initWithRootView:rootView splashScreenView:splashScreenView];
+  
+  if (self) {
```
Comment (bbarthec) [inline ios/Exponent/Kernel/Views/Loading/EXManagedAppSplashScreenController.m:28]:
```suggestion
  if (self = [super initWithRootView:rootView splashScreenView:splashScreenView]) {
```
This is a common pattern in Objective-C to inline `if (self = [super ...])` 😉

File: ios/Exponent/Kernel/Views/Loading/EXManagedAppSplashScreenController.m [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+  self.warningTimer = [NSTimer scheduledTimerWithTimeInterval:20.0
+                                                       target:self
+                                                     selector:@selector(showSplashScreenVisibleWarning)
+                                                     userInfo:nil
+                                                      repeats:NO];
+}
+
+- (void)showSplashScreenVisibleWarning
+{
+#if DEBUG
```
Comment (bbarthec) [inline ios/Exponent/Kernel/Views/Loading/EXManagedAppSplashScreenController.m:71]: Don't we want to behave the same way on Android? I've failed to find `if (debug)` there for `startSplashScreenWarningTimer` method in `ManagedAppSplashScreenController.kt` file  🤔

**Review by bbarthec — CHANGES_REQUESTED** [review]
Overall it looks fine, but I've got a few critical suggestions:
- don't litter `splashScreenViewProvider`'s with view-unrelated methods (single responsibility principle in this case, but provider name does suggest there's nothing view-unrelated inside it), move them somewhere else, eg. to `SplashScreen(View)Controllers`?
- pay attention to the current codestyle. I know our Objective-C codestyle is not perfect, but my opinion is that having consistency is better than dealing with multiple different codestyles (in the first case I keep my focus on the code, and in the latter I'm being distracted by the inconsistencies in the code style 🤷) 
- some other things left in the code.

**Comment by ajsmth** [comment]
hey @bbarthec  - I've updating the implementation - hopefully its closer to what you had in mind! Thanks for all of the feedback :)

File: ios/Exponent/Kernel/Views/Loading/EXManagedAppSplashScreenController.h [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+//  Created by andrew on 2021-06-07.
+//  Copyright © 2021 650 Industries. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+#import <EXSplashScreen/EXSplashScreenViewController.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface EXManagedAppSplashScreenController : EXSplashScreenViewController
```
Comment (bbarthec) [inline ios/Exponent/Kernel/Views/Loading/EXManagedAppSplashScreenController.h:14]: As we're having `EXSplashScreenViewController`, let's also have:
```suggestion
@interface EXManagedAppSplashScreenViewController : EXSplashScreenViewController
```

File: packages/expo-splash-screen/android/src/main/java/expo/modules/splashscreen/singletons/SplashScreen.kt [hunk 23]
```diff
@@ -82,6 +75,38 @@ object SplashScreen : SingletonModule {
...
+   * @throws [expo.modules.splashscreen.exceptions.NoContentViewException] when [SplashScreen.show] is called before [Activity.setContentView] (when no ContentView is present for given activity).
+   */
+  @JvmStatic
+  @JvmOverloads
+  fun show(
+          activity: Activity,
+          splashScreenViewController: SplashScreenViewController,
+          statusBarTranslucent: Boolean,
+          successCallback: () -> Unit = {},
+          failureCallback: (reason: String) -> Unit = { Log.w(TAG, it) }
```
Comment (bbarthec) [inline packages/expo-splash-screen/android/src/main/java/expo/modules/splashscreen/singletons/SplashScreen.kt:97]:
```suggestion
  activity: Activity,
  splashScreenViewController: SplashScreenViewController,
  statusBarTranslucent: Boolean,
  successCallback: () -> Unit = {},
  failureCallback: (reason: String) -> Unit = { Log.w(TAG, it) }
```

**Review by bbarthec — APPROVED** [review]
